File: RUSH150/Construct Binary Tree from Preorder and Inorder Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    if (!preorder.length > 0){return null}

    const root = new TreeNode( val = preorder[0] ) ;
    const indexOfRoot = inorder.indexOf(root.val) ; 
    // indexOfRoot = how many node in left sub-tree , and the rest is belong to right sub-tree 
    const leftTree = buildTree( preorder.slice( 1 , 1+indexOfRoot ) , inorder.slice(0 ,indexOfRoot)   )

    const rightTree = buildTree( preorder.slice(1+indexOfRoot) , inorder.slice( indexOfRoot+1 ) )

    root.left = leftTree
    root.right = rightTree 

    return root 

};

File: RUSH/Construct Binary Tree from Preorder and Inorder Traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

from typing import List , Optional

class Solution:
    
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        
        
        def build(preorder:list , inorder:list):  
            
            if len(preorder) == 0 : return None  
            
            rootIndex = inorder.index(  preorder[0]  )
            
            root = TreeNode(val = inorder[rootIndex]) 
            left_sub_tree = build( preorder[1:rootIndex+1]   , inorder[:rootIndex]   )
            right_sub_tree = build(  preorder[rootIndex+1:] , inorder[rootIndex+1:]  )
            
            root.left = left_sub_tree 
            root.right = right_sub_tree 
            
            return root 

        return build(preorder , inorder)